File: 2020/day-8/aoc-8.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Dec  8 05:47:51 2020

@author: mattiaruefenacht
"""

import os

os.getcwd()

os.chdir(r'/Users/mattiaruefenacht/Desktop/AOC2020/day-8')

def hasInfiniteLoop(data):
    visited = set()
    accumulated = 0
    counter = 0

    while True:
        if counter == len(data) - 1:
            return (False, accumulated)

        instruction = data[counter]

        if counter in visited:
            return (True, accumulated)
        elif instruction[0] == "acc":
            accumulated += int(instruction[1])
        elif instruction[0] == "jmp":
            counter += int(instruction[1])
            continue

        visited.add(counter)
        counter += 1

def findWrongInstruction(data):
    for ind, instr in enumerate(data):
        if instr[0] == "jmp" or instr[0] == "nop":
            data_copy = data.copy()
            new_instr = "nop" if instr[0] == "jmp" else "jmp"
            data_copy[ind] = [new_instr, data[ind][1]]

            test = hasInfiniteLoop(data_copy)
            if not test[0]:
                return test[1]

def main():
    doc = open("input.txt", "r")
    data = [x.strip().split() for x in doc.read().strip().split("\n")]

    print('Part 1 solution:', hasInfiniteLoop(data)[1])
    print('Part 2 solution:', findWrongInstruction(data))

if __name__ == "__main__":
    main()

File: 2020/day-8/aoc-8.go
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type instr struct {
	Op  string
	Arg int
}
type instructions []instr

func run(lines instructions) (int, error) {
	accumulator := 0
	seen := make(map[int]bool)
	for i := 0; i < len(lines); i++ {
		_, ran := seen[i]
		if ran {
			return accumulator, errors.New("Already ran.")
		}
		seen[i] = true

		op := lines[i].Op
		arg := lines[i].Arg
		switch op {
		case "jmp":
			i += arg - 1
		case "acc":
			accumulator += arg
		}
	}
	return accumulator, nil
}

func main() {
	data, err := ioutil.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	contents := string(data)
	split := strings.Split(contents, "\n")

	bootcode := make(instructions, 0)
	for i := 0; i < len(split)-1; i++ {
		parts := strings.Fields(split[i])
		arg, _ := strconv.Atoi(parts[1])
		bootcode = append(bootcode, instr{Op: parts[0], Arg: arg})
	}

	accumulator, err := run(bootcode)
	if err != nil {
		fmt.Println("Part 1 solution:", accumulator)
	}

	for i := 0; i <= len(bootcode); i++ {
		op := bootcode[i].Op
		switch op {
		case "jmp":
			bootcode[i].Op = "nop"
		case "nop":
			bootcode[i].Op = "jmp"
		}
		accumulator, err := run(bootcode)
		bootcode[i].Op = op
		if err == nil {
			fmt.Println("Part 2 solution:", accumulator)
			break
		}
	}
}
